fix sanity test failures

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->


**What happened**:

```
Summarizing 5 Failures:

[Fail] Controller Service CreateVolume [It] should not fail when requesting to create a volume with already existing name and same capacity. 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:371

[Fail] Controller Service CreateVolume [It] should fail when requesting to create a volume with already existing name and different capacity. 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:442

[Fail] Controller Service ValidateVolumeCapabilities [It] should fail when the requested volume does not exist 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:705

[Fail] Node Service [It] should work 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/node.go:464

Ran 26 of 57 Specs in 77.639 seconds
FAIL! -- 21 Passed | 5 Failed | 0 Pending | 31 Skipped
```

following issue has been fixed:
```
[Fail] Controller Service DeleteVolume [It] should succeed when an invalid volume id is used 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:543
```
**What you expected to happen**:

**How to reproduce it**:
run sanity test: https://github.com/kubernetes-sigs/azurefile-csi-driver/tree/master/test/sanity

**Anything else we need to know?**:

**Environment**:
- CSI Driver version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

Update: there are 4 sanity test failures originally:
```
Summarizing 4 Failures:
[Fail] Controller Service CreateVolume [It] should not fail when requesting to create a volume with already existing name and same capacity. 
/home/travis/gopath/pkg/mod/github.com/kubernetes-csi/csi-test@v1.1.0/pkg/sanity/controller.go:371
[Fail] Controller Service CreateVolume [It] should fail when requesting to create a volume with already existing name and different capacity. 
/home/travis/gopath/pkg/mod/github.com/kubernetes-csi/csi-test@v1.1.0/pkg/sanity/controller.go:442
[Fail] Controller Service ValidateVolumeCapabilities [It] should fail when the requested volume does not exist 
/home/travis/gopath/pkg/mod/github.com/kubernetes-csi/csi-test@v1.1.0/pkg/sanity/controller.go:705
[Fail] Node Service [It] should work 
/home/travis/gopath/pkg/mod/github.com/kubernetes-csi/csi-test@v1.1.0/pkg/sanity/node.go:464
Ran 26 of 57 Specs in 21.016 seconds
FAIL! -- 22 Passed | 4 Failed | 0 Pending | 31 Skipped
--- FAIL: TestSanity (21.06s)
```

--- a/pkg/azurefile/azurefile.go
+++ b/pkg/azurefile/azurefile.go
@@ -21,9 +21,12 @@ import (
 	"strings"
 
 	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume/util"
 
+	azs "github.com/Azure/azure-sdk-for-go/storage"
 	"github.com/container-storage-interface/spec/lib/go/csi"
 	csicommon "github.com/kubernetes-sigs/azurefile-csi-driver/pkg/csi-common"
+	"github.com/pborman/uuid"
 	"k8s.io/klog"
 	"k8s.io/legacy-cloud-providers/azure"
 )
@@ -38,6 +41,10 @@ const (
 	defaultFileMode  = "0777"
 	defaultDirMode   = "0777"
 	defaultVers      = "3.0"
+
+	// See https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-shares--directories--files--and-metadata#share-names
+	fileShareNameMinLength = 3
+	fileShareNameMaxLength = 63
 )
 
 // Driver implements all interfaces of CSI drivers
@@ -99,6 +106,33 @@ func (d *Driver) Run(endpoint string) {
 	// Driver d act as IdentityServer, ControllerServer and NodeServer
 	s.Start(endpoint, d, d, d)
 	s.Wait()
+}
+
+func (d *Driver) checkFileShareExists(accountName, resourceGroup, name string) (bool, error) {
+	// find the access key with this account
+	accountKey, err := d.cloud.GetStorageAccesskey(accountName, resourceGroup)
+	if err != nil {
+		// Because the account does not exist in the resource group, the file share could not exist.
+		// In the next step, we will create a temp account to create file share.
+		// So here we return nil, and warn the user.
+		klog.Warningf("could not get storage key for storage account %s: %v", accountName, err)
+		return false, nil
+	}
+
+	fileClient, err := d.getFileSvcClient(accountName, accountKey)
+	if err != nil {
+		return false, fmt.Errorf("error creating azure client: %v", err)
+	}
+	return fileClient.GetShareReference(name).Exists()
+}
+
+func (d *Driver) getFileSvcClient(accountName, accountKey string) (*azs.FileServiceClient, error) {
+	fileClient, err := azs.NewClient(accountName, accountKey, d.cloud.Environment.StorageEndpointSuffix, azs.DefaultAPIVersion, true)
+	if err != nil {
+		return nil, err
+	}
+	fc := fileClient.GetFileService()
+	return &fc, nil
 }
 
 // get file share info according to volume id, e.g.
@@ -179,4 +213,36 @@ func getStorageAccount(secrets map[string]string) (string, string, error) {
 	}
 
 	return accountName, accountKey, nil
+}
+
+// File share names can contain only lowercase letters, numbers, and hyphens,
+// and must begin and end with a letter or a number,
+// and must be from 3 through 63 characters long.
+// The name cannot contain two consecutive hyphens.
+//
+// See https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-shares--directories--files--and-metadata#share-names
+func getValidFileShareName(volumeName string) string {
+	fileShareName := strings.ToLower(volumeName)
+	if len(fileShareName) > fileShareNameMaxLength {
+		fileShareName = fileShareName[0:fileShareNameMaxLength]
+	}
+	if !checkShareNameBeginAndEnd(fileShareName) || len(fileShareName) < fileShareNameMinLength {
+		fileShareName = util.GenerateVolumeName("pvc-file", uuid.NewUUID().String(), fileShareNameMaxLength)
+		klog.Warningf("the requested volume name (%q) is invalid, so it is regenerated as (%q)", volumeName, fileShareName)
+	}
+	fileShareName = strings.Replace(fileShareName, "--", "-", -1)
+
+	return fileShareName
+}
+
+func checkShareNameBeginAndEnd(fileShareName string) bool {
+	length := len(fileShareName)
+	if (('a' <= fileShareName[0] && fileShareName[0] <= 'z') ||
+		('0' <= fileShareName[0] && fileShareName[0] <= '9')) &&
+		(('a' <= fileShareName[length-1] && fileShareName[length-1] <= 'z') ||
+			('0' <= fileShareName[length-1] && fileShareName[length-1] <= '9')) {
+		return true
+	}
+
+	return false
 }

--- a/pkg/azurefile/azurefile_test.go
+++ b/pkg/azurefile/azurefile_test.go
@@ -214,3 +214,77 @@ func TestGetStorageAccount(t *testing.T) {
 		}
 	}
 }
+
+func TestGetValidFileShareName(t *testing.T) {
+	tests := []struct {
+		volumeName string
+		expected   string
+	}{
+		{
+			volumeName: "aqz",
+			expected:   "aqz",
+		},
+		{
+			volumeName: "029",
+			expected:   "029",
+		},
+		{
+			volumeName: "a--z",
+			expected:   "a-z",
+		},
+		{
+			volumeName: "A2Z",
+			expected:   "a2z",
+		},
+		{
+			volumeName: "1234567891234567891234567891234567891234567891234567891234567891",
+			expected:   "123456789123456789123456789123456789123456789123456789123456789",
+		},
+	}
+
+	for _, test := range tests {
+		result := getValidFileShareName(test.volumeName)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("input: %q, getValidFileShareName result: %q, expected: %q", test.volumeName, result, test.expected)
+		}
+	}
+}
+
+func TestCheckShareNameBeginAndEnd(t *testing.T) {
+	tests := []struct {
+		fileShareName string
+		expected      bool
+	}{
+		{
+			fileShareName: "aqz",
+			expected:      true,
+		},
+		{
+			fileShareName: "029",
+			expected:      true,
+		},
+		{
+			fileShareName: "a-9",
+			expected:      true,
+		},
+		{
+			fileShareName: "0-z",
+			expected:      true,
+		},
+		{
+			fileShareName: "-1-",
+			expected:      false,
+		},
+		{
+			fileShareName: ":1p",
+			expected:      false,
+		},
+	}
+
+	for _, test := range tests {
+		result := checkShareNameBeginAndEnd(test.fileShareName)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("input: %q, checkShareNameBeginAndEnd result: %v, expected: %v", test.fileShareName, result, test.expected)
+		}
+	}
+}

--- a/pkg/azurefile/controllerserver.go
+++ b/pkg/azurefile/controllerserver.go
@@ -24,10 +24,10 @@ import (
 	volumehelper "github.com/kubernetes-sigs/azurefile-csi-driver/pkg/util"
 
 	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2018-07-01/storage"
+
 	"github.com/container-storage-interface/spec/lib/go/csi"
-	"github.com/pborman/uuid"
+
 	"k8s.io/klog"
-	"k8s.io/kubernetes/pkg/volume/util"
 
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
@@ -52,8 +52,10 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 	volSizeBytes := int64(req.GetCapacityRange().GetRequiredBytes())
 	requestGiB := int(volumehelper.RoundUpGiB(volSizeBytes))
 
+	capacityBytes := req.GetCapacityRange().GetRequiredBytes()
+
 	parameters := req.GetParameters()
-	var sku, resourceGroup, location, account, fileShareName string
+	var sku, resourceGroup, location, account, fileShareName, volumeID string
 
 	// Apply ProvisionerParameters (case-insensitive). We leave validation of
 	// the values to the cloud provider.
@@ -83,31 +85,35 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 	}
 
 	if fileShareName == "" {
-		// File share name has a length limit of 63, and it cannot contain two consecutive '-'s.
-		fileShareName = util.GenerateVolumeName("pvc-file", uuid.NewUUID().String(), 63)
-		fileShareName = strings.Replace(fileShareName, "--", "-", -1)
+		fileShareName = getValidFileShareName(name)
 	}
 
-	klog.V(2).Infof("begin to create file share(%s) on account(%s) type(%s) rg(%s) location(%s) size(%d)", fileShareName, account, sku, resourceGroup, location, requestGiB)
-	retAccount, _, err := d.cloud.CreateFileShare(fileShareName, account, sku, accountKind, resourceGroup, location, requestGiB)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create file share(%s) on account(%s) type(%s) rg(%s) location(%s) size(%d), error: %v", fileShareName, account, sku, resourceGroup, location, requestGiB, err)
-	}
-	volumeID := fmt.Sprintf(volumeIDTemplate, resourceGroup, retAccount, fileShareName)
-
-	/* todo: snapshot support
-	if req.GetVolumeContentSource() != nil {
-		contentSource := req.GetVolumeContentSource()
-		if contentSource.GetSnapshot() != nil {
+	if exists, err := d.checkFileShareExists(account, resourceGroup, fileShareName); err != nil {
+		return nil, err
+	} else if exists {
+		volumeID = fmt.Sprintf(volumeIDTemplate, resourceGroup, account, fileShareName)
+	} else {
+		klog.V(2).Infof("begin to create file share(%s) on account(%s) type(%s) rg(%s) location(%s) size(%d)", fileShareName, account, sku, resourceGroup, location, requestGiB)
+		retAccount, _, err := d.cloud.CreateFileShare(fileShareName, account, sku, accountKind, resourceGroup, location, requestGiB)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create file share(%s) on account(%s) type(%s) rg(%s) location(%s) size(%d), error: %v", fileShareName, account, sku, resourceGroup, location, requestGiB, err)
 		}
+		volumeID = fmt.Sprintf(volumeIDTemplate, resourceGroup, retAccount, fileShareName)
+
+		/* todo: snapshot support
+		if req.GetVolumeContentSource() != nil {
+			contentSource := req.GetVolumeContentSource()
+			if contentSource.GetSnapshot() != nil {
+			}
+		}
+		*/
+		klog.V(2).Infof("create file share %s on storage account %s successfully", fileShareName, retAccount)
 	}
-	*/
-	klog.V(2).Infof("create file share %s on storage account %s successfully", fileShareName, retAccount)
 
 	return &csi.CreateVolumeResponse{
 		Volume: &csi.Volume{
 			VolumeId:      volumeID,
-			CapacityBytes: req.GetCapacityRange().GetRequiredBytes(),
+			CapacityBytes: capacityBytes,
 			VolumeContext: parameters,
 		},
 	}, nil

